jaxrs-clients/src/main/java/com/palantir/remoting3/jaxrs/RemoteIoExceptionClient.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.jaxrs;
+
+import com.palantir.remoting3.okhttp.RemoteIoException;
+import feign.Client;
+import feign.Request;
+import feign.Response;
+import java.io.IOException;
+
+/**
+ * This client's sole purpose is to turn our special {@link RemoteIoException} into a RuntimeException.
+ * <p>
+ * It is necessary because okhttp only propagates IOExceptions nicely through all its async futures {@see
+ * okhttp3.RealCall}, but unfortunately feign turns IOExceptions into ugly FeignExceptions in its {@link
+ * feign.SynchronousMethodHandler}.
+ */
+public final class RemoteIoExceptionClient implements Client {
+    private final Client delegate;
+
+    public RemoteIoExceptionClient(Client delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Response execute(Request request, Request.Options options) throws IOException {
+        try {
+            return delegate.execute(request, options);
+        } catch (RemoteIoException e) {
+            throw e.getRuntimeExceptionCause();
```

Comment:
as much as possible I'd prefer we don't blindly rethrow exceptions; can we do something here to make sure that the stack trace is the right place. Can we leverage the fact we know that this is a RemoteException and wrap it maybe with another identical RemoteException?

Reply:
OK so just to confirm I'm understanding: re-throwing the internal 'cause' would lead to confusing stacktraces because it would would omit lines for all this RemoteIoException handling stuff and just point at the original constuction location... you'd prefer that we actually construct a new exception here with the same semantic purpose.